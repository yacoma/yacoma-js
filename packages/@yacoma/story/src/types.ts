import { IConfiguration, Overmind, IOperator } from 'overmind'

type PropsFunction<Config extends IConfiguration, Props> = (
  app: Overmind<Config>
) => Props

export interface TStory<Config extends IConfiguration = any, Props = any> {
  // The name of the story
  name: string
  // What is the proper type here ?
  // If not set here, it must be set in the Stories defaults.
  component?: any
  // A component to wrap the tested component.
  wrapper?: any
  // A sequence triggered on component wrapper click
  titleClick?: IOperator<Config, React.MouseEvent, React.MouseEvent>
  // Turn on debugger for this story by providing a port number (should be unique
  // for each story).
  devtools?: string
  // Regular props passed to component.
  props?: Props | PropsFunction<Config, Props>
  state?: any
  actions?: {
    [key: string]: any
  }
  config?: Config
  // theme override
  theme?: {
    [key: string]: any
  }
  children?: any
  // jsx to use this story (use this if there are children and/or the guessing of
  // component name does not work)
  jsx?: string
}
export type Story<Props = any> = TStory<any, Props>

export interface TStories<Config extends IConfiguration = any, Props = any> {
  name: string
  stories: TStory<Config, Props>[]
  // Defaults for all stories in this set
  component?: any
  wrapper?: any
  titleClick?: IOperator<Config, React.MouseEvent, React.MouseEvent>
  // Regular props passed to component.
  props?: Props | PropsFunction<Config, Props>
  state?: any
  actions?: {
    [key: string]: any
  }
  config?: Config
  theme?: {
    [key: string]: any
  }
  children?: any
}
export type Stories<Props = any> = TStories<any, Props>

export interface StoryConfig {
  state: {
    stories: {}
  }
}
