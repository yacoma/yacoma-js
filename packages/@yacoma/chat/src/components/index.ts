export * from './Attachment'
export * from './CollectionMessages'
export * from './ConfirmShareAttachments'
export * from './ContactIcon'
export * from './ContactTitle'
export * from './Explorer'
export * from './MessageInProgress'
export * from './Messages'
export * from './NewContactIcon'
export * from './People'
export * from './SearchIcon'
