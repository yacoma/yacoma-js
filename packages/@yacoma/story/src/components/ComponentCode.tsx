import * as React from 'react'
import { Comp, styled, useOvermind } from '../app'
import { stringifyComponent } from '../helpers/stringifyComponent'

function limit(text: string, length: number) {
  const lines = text.split('\n')
  const line = lines[0]
  if (line.length <= length && lines.length === 1) {
    return line
  } else {
    return line.slice(0, length) + '...'
  }
}

interface CodeProps {
  props: Function | { [key: string]: any }
  component: any
  className?: string
  jsx?: string
}

const Wrapper = styled.div`
  background: #e2e2e2;
  color: #555;
  padding: 10px;
  border-right: 1px solid #aaa;
  font-size: 0.8rem;
  white-space: pre;
  position: relative;
  width: 15rem;
  & .jsx {
    border: 1px solid #aaa;
    background: #f5f4de;
    position: absolute;
    padding: 10px;
    top: -1px;
    left: -1px;
    opacity: 0;
    transition-property: opacity;
    transition-duration: 0.1s;
    transition-delay: 0.8s;
    z-index: 99;
    box-shadow: 4px 4px 10px rgba(0, 0, 0, 0.3);
    visibility: hidden;
    min-width: 15rem;
  }
  &:hover .jsx {
    visibility: visible;
    opacity: 1;
  }
`

export const ComponentCode: Comp<CodeProps> = ({
  props,
  children,
  component,
  className,
  jsx,
}) => {
  const app = useOvermind()
  const theProps = typeof props === 'function' ? props(app) : props
  return (
    <Wrapper className={className}>
      {limit(
        jsx || stringifyComponent(component, { ...theProps, children }),
        30
      )}
      <div className="jsx">
        {jsx || stringifyComponent(component, { ...theProps, children })}
      </div>
    </Wrapper>
  )
}
