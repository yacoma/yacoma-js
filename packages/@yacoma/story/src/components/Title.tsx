import classnames from 'classnames'
import * as React from 'react'
import { Comp, styled } from '../app'

interface TitleProps {
  className?: string
  onClick?: () => void
}

const ATitle = styled.div`
  flex-grow: 1;
  background: #e2e2e2;
  padding: 10px;
  font-size: 0.9rem;
  width: 15rem;
  user-select: none;
  border-right: 1px solid #aaa;
`

export const Title: Comp<TitleProps> = ({ children, className, onClick }) => (
  <ATitle
    className={classnames('Title', className)}
    onClick={onClick}
    style={{ cursor: onClick ? 'pointer' : 'inherit' }}
  >
    {children}
  </ATitle>
)
