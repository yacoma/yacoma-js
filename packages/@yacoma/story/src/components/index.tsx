export { StoryRouter } from './StoryRouter'
export { ComponentWrapper } from './StoryDisplay'
