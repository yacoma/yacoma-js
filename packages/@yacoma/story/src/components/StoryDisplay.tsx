import { IConfiguration, IOperator, Overmind } from 'overmind'
import { Provider } from 'overmind-react'
import * as React from 'react'
import { DirectThemeProvider, themeProxy } from '@yacoma/theme'
import { Comp, styled, useOvermind } from '../app'
import { storyTheme } from '../theme'
import { Stories, Story } from '../types'
import { ComponentCode } from './ComponentCode'
import { Title } from './Title'

const theme = themeProxy(storyTheme)

export const ComponentWrapper = styled.div`
  align-self: end;
  margin: ${theme.storyGridSize};
  background: white;
`

const Wrapper = styled.div`
  display: flex;
  flex-direction: row;
  background-size: calc(1 * ${theme.storyGridSize})
    calc(1 * ${theme.storyGridSize});
  background-position: calc(1 * ${theme.storyGridSize})
    calc(1 * ${theme.storyGridSize});
  background-image: linear-gradient(
      to right,
      ${theme.storyGridColor} 1px,
      transparent 1px
    ),
    linear-gradient(
      to bottom,
      ${theme.storyGridColor} 1px,
      ${theme.storyGridBackground} 1px
    );
`

const ColWrap = styled.div`
  display: flex;
  flex-direction: column;
`

export interface StoryDisplayProps {
  className?: string
  story: Story
  stories: Stories
}

const Display: Comp<{ getProps: any; Component: any }> = ({
  children,
  getProps,
  Component,
}) => {
  const app = useOvermind()
  const props = getProps(app)
  return <Component {...props}>{children}</Component>
}

function deepMerge(
  target: { [key: string]: any },
  source: { [key: string]: any },
  keys: string[] = []
) {
  Object.keys(source).forEach(key => {
    const value = source[key]
    if (Array.isArray(value)) {
      target[key] = value
    } else if (value === null) {
      delete target[key]
    } else if (typeof value === 'object') {
      target[key] =
        target[key] === undefined ? {} : Object.assign({}, target[key])
      if (typeof target[key] === 'object') {
        deepMerge(target[key], value, [...keys, key])
      } else {
        throw new Error(
          `Cannot merge object in '${[...keys, key].join(
            '.'
          )}: target is not an object.'`
        )
      }
    } else if (typeof target[key] === 'object') {
      throw new Error(
        `Cannot merge ${typeof value} in '${[...keys, key].join(
          '.'
        )}: target is an object.'`
      )
    } else {
      target[key] = source[key]
    }
  })
}

function StoryDisplayComp({ className, story, stories }: StoryDisplayProps) {
  function getProps(app: any) {
    const props = stories.props || {}
    const sprops = story.props || {}
    return Object.assign(
      {},
      typeof props === 'function' ? props(app) : props,
      typeof sprops === 'function' ? sprops(app) : sprops
    )
  }
  const Comp = story.component || stories.component
  if (!Comp) {
    throw new Error(
      `Invalid story '${story.name}' in '${stories.name}': missing component.`
    )
  }

  const appState = Object.assign(
    {},
    (story.config || stories.config || {}).state || ({} as any)
  )
  if (stories.state) {
    deepMerge(appState, stories.state)
  }
  if (story.state) {
    deepMerge(appState, story.state)
  }

  const config: IConfiguration & {
    actions: { _titleClick?: IOperator<any, any, any> }
    state: {
      theme?: any
    }
  } = Object.assign(
    {
      actions: {},
    },
    story.config || stories.config || {},
    {
      state: appState,
    }
  )

  const titleClick = story.titleClick || stories.titleClick
  if (titleClick) {
    config.actions = Object.assign({}, config.actions, {
      _titleClick: titleClick,
    })
  }

  const name = `${stories.name}.${story.name}`

  const overmind = new Overmind(config, {
    name,
    devtools: story.devtools,
  })

  const CompWrap = story.wrapper || stories.wrapper || ComponentWrapper

  const TheChildren = story.children || stories.children || null
  let children = TheChildren
  if (TheChildren && typeof TheChildren === 'function') {
    children = <TheChildren />
  }
  const mtheme = overmind.state.theme
    ? overmind.state.theme.selectedTheme
    : undefined
  const theme = Object.assign(
    {},
    mtheme || {},
    storyTheme,
    stories.theme || {},
    story.theme || {}
  )
  const scopeName = `${stories.name}-${story.name}`.replace(/[^a-zA-Z]/g, '-')
  return (
    <DirectThemeProvider scopeName={scopeName} theme={theme}>
      <Provider value={overmind}>
        <div className={className}>
          <Wrapper>
            <ColWrap>
              <Title onClick={overmind.actions._titleClick}>{story.name}</Title>
              <ComponentCode
                component={Comp}
                children={children}
                props={getProps}
                jsx={story.jsx}
              />
            </ColWrap>
            <CompWrap>
              <Display
                Component={Comp}
                getProps={getProps}
                children={children}
              />
            </CompWrap>
          </Wrapper>
        </div>
      </Provider>
    </DirectThemeProvider>
  )
}

export const StoryDisplay = styled(StoryDisplayComp)`
  border: 1px solid #aaa;
  margin: 10px;
  display: flex;
  flex-direction: column;
`
