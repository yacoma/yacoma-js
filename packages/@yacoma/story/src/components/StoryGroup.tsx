import * as React from 'react'
import { useState } from 'react'
import styled from 'styled-components'
import { Comp } from '../app'
import { Stories } from '../types'
import { StoryDisplay } from './StoryDisplay'

const GroupTitle = styled.div`
  padding: 10px;
  cursor: pointer;
  user-select: none;
`

const GroupStories = styled.div``

interface StoryGroupProps {
  className?: string
  stories: Stories
}

const StoryGroupComp: Comp<StoryGroupProps> = ({ className, stories }) => {
  const [open, setOpen] = useState(true)

  /* istanbul ignore next */
  const style = { display: open ? 'block' : 'none' }
  return (
    <div className={className}>
      <GroupTitle onClick={/* istanbul ignore next */ () => setOpen(!open)}>
        {stories.name}
      </GroupTitle>
      <GroupStories style={style}>
        {stories.stories.map((story, idx) => (
          <StoryDisplay key={idx} story={story} stories={stories} />
        ))}
      </GroupStories>
    </div>
  )
}

export const StoryGroup = styled(StoryGroupComp)`
  border-left: 3px solid #eee;
  margin: 10px 0;
  display: flex;
  flex-direction: column;
`
