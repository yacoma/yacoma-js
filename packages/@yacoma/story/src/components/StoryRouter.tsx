import * as React from 'react'
import { Comp, styled } from '../app'
import { Stories } from '../types'
import { StoryGroup } from './StoryGroup'

interface StoryRouterProps {
  className?: string
  stories: Stories[]
}

const StoryRouterComp: Comp<StoryRouterProps> = ({ className, stories }) => (
  <div className={className}>
    {stories.map((group, idx) => (
      <StoryGroup key={idx} stories={group} />
    ))}
  </div>
)

export const StoryRouter = styled(StoryRouterComp)`
  display: flex;
  flex-direction: column;
  font-family: Consolas, Menlo, Monaco, Lucida Console, Liberation Mono,
    DejaVu Sans Mono, Bitstream Vera Sans Mono, monospace;
`
