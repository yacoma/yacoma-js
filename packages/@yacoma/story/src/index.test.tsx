import * as React from 'react'
import renderer from 'react-test-renderer'
import simple from 'simple-mock'
import { makeApp } from './'
import { StoryDisplay } from './components/StoryDisplay'
import { Stories, Story } from './types'

const component = ({ name }: { name: string }) => <div>This is {name}</div>

const NoProps: Story = {
  name: 'NoProps',
  component,
  titleClick: action => action.run(() => {}),
}

const SimpleProps: Story = {
  name: 'Simple',
  component,
  props: { name: 'foo' },
}

const WithClick: Story = {
  name: 'withClick',
  component,
  titleClick: action => action.run(() => {}),
  props: { name: 'foo' },
}

const stories: Stories = {
  name: 'Foo',
  stories: [NoProps, SimpleProps, WithClick],
}

describe('stories', () => {
  it('should render stories without error', () => {
    const App = makeApp({}, { stories })
    const tree = renderer.create(<App />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

interface ErrorProps {
  errors: string[]
}

class ErrorCatcher extends React.Component<ErrorProps> {
  constructor(props: ErrorProps) {
    super(props)
  }

  componentDidCatch(error: Error) {
    this.props.errors.push(error.message)
  }

  render() {
    return this.props.children
  }
}

it('should thrown on missing component', () => {
  const bang: Story = {
    name: 'Bang',
  }
  const stories: Stories = {
    name: 'Fail stories',
    stories: [],
  }
  const errors: string[] = []
  simple.mock(console, 'error').callFn(() => {})
  renderer.create(
    <ErrorCatcher errors={errors}>
      <StoryDisplay stories={stories} story={bang} />
    </ErrorCatcher>
  )
  expect(errors).toEqual([
    "Invalid story 'Bang' in 'Fail stories': missing component.",
  ])
  simple.restore()
})
