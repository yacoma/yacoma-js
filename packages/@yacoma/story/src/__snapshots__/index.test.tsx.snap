// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories should render stories without error 1`] = `
<div
  className="sc-htoDjs hsuxup"
>
  <div
    className="sc-gzVnrw fJrtRC"
  >
    <div
      className="sc-EHOje klrYJJ"
      onClick={[Function]}
    >
      Foo
    </div>
    <div
      className="sc-bZQynM bOrPFB"
      style={
        Object {
          "display": "block",
        }
      }
    >
      <div
        className="theme-Foo-NoProps"
      >
        <div
          className="sc-ifAKCX hYWxfb"
        >
          <div
            className="sc-htpNat hIwNRt"
          >
            <div
              className="sc-bxivhb glCqXO"
            >
              <div
                className="Title sc-bdVaJa dLCEWS"
                onClick={[Function]}
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
              >
                NoProps
              </div>
              <div
                className="sc-dnqmqq jxrIyk"
              >
                &lt;component /&gt;
                <div
                  className="jsx"
                >
                  &lt;component /&gt;
                </div>
              </div>
            </div>
            <div
              className="sc-bwzfXH dNRbgB"
            >
              <div>
                This is 
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="theme-Foo-Simple"
      >
        <div
          className="sc-ifAKCX hYWxfb"
        >
          <div
            className="sc-htpNat hIwNRt"
          >
            <div
              className="sc-bxivhb glCqXO"
            >
              <div
                className="Title sc-bdVaJa dLCEWS"
                style={
                  Object {
                    "cursor": "inherit",
                  }
                }
              >
                Simple
              </div>
              <div
                className="sc-iwsKbI dAfjtB"
              >
                &lt;component name="foo" /&gt;
                <div
                  className="jsx"
                >
                  &lt;component name="foo" /&gt;
                </div>
              </div>
            </div>
            <div
              className="sc-bwzfXH dNRbgB"
            >
              <div>
                This is 
                foo
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="theme-Foo-withClick"
      >
        <div
          className="sc-ifAKCX hYWxfb"
        >
          <div
            className="sc-htpNat hIwNRt"
          >
            <div
              className="sc-bxivhb glCqXO"
            >
              <div
                className="Title sc-bdVaJa dLCEWS"
                onClick={[Function]}
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
              >
                withClick
              </div>
              <div
                className="sc-gZMcBi jGsjoc"
              >
                &lt;component name="foo" /&gt;
                <div
                  className="jsx"
                >
                  &lt;component name="foo" /&gt;
                </div>
              </div>
            </div>
            <div
              className="sc-bwzfXH dNRbgB"
            >
              <div>
                This is 
                foo
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
